0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,77 @@
+# Pre-Shared Key
+
+The Pre-Shared Key (PSK) protocol is an end-to-end transport protocol, used by the sender and receiver of an ILP payment to decide on a condition and fulfillment for a payment. The protocol also encrypts any additional data sent along with the payment. The encryption algorithm used is AES-256-CTR.
+
+As the name suggests, PSK relies on a pre-shared secret key that is shared beforehand. In application-layer protocols like SPSP, the receiver generates the shared key and shares it with the sender over an end-to-end secure communication channel such as HTTPS. 
+
+An advantage of PSK is that the pre-shared secret key only needs to be shared once. So long as both parties possess this key and are listening for transfers, they can send payments between one another.
+
+A disadvantage of PSK is that it is repudiable. Knowledge of a payment's fulfillment cannot be used as cryptographic proof that the receiver completed the payment, because both sender and receiver know how to generate the fulfillment with their pre-shared secret key. This causes no issues unless you have some other framework on top of ILP that relies on this cryptographic proof.
+
+## Flow
```

Comment:
You could reorder this flow. This is somewhat related to the outer/inner -> public/private nomenclature comment below.

The general idea is to talk about the data in the order that it appears (changes in bold):

1. The pre-shared secret key is shared out of band.
2. The sender creates a random token value, and calculates an HMAC of the token using the pre-shared secret key as the key. This HMAC will be referred to as the payment key.
3. **The sender constructs the PSK data**
   1. **The sender emits the PSK version line `PSK/1.0\n`**
   2. **The sender emits the public headers (containing the token), followed by `\n\n`**
   3. **The remainder of the PSK data after this point will be encrypted using AES-256-CTR with a key derived from the payment key**

      **Note: The ciphertext is not base64-encoded.**
   4. **The sender emits the private headers, followed by `\n\n`**
   5. **The sender emits the application data.**
4. The sender quotes to the receiver and sends a transfer **with the PSK data** in the data field of the ILP packet. The fulfillment of this payment is the HMAC of the ILP packet using the payment key as the key.
5. The receiver gets notified by their ledger of an incoming prepared transfer with an ILP packet.
6. The receiver generates the payment key **using the token** contained in the **public headers**, and uses it to decrypt the encrypted portion of the PSK data.
7. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. 8. The receiver may also call a backend to make sure the incoming funds are expected.
9. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key.


(Note: I changed some of the wording to make it clear that the encryption key is **not** the payment key, but a key derived from it. We should explain precisely how all of the keys are derived. For more information on why we don't want to just use the payment key for both encryption and MAC, please [see here](https://security.stackexchange.com/questions/37880/why-cant-i-use-the-same-key-for-encryption-and-mac). In my initial implementation, I actually derived one key for encryption and one for the HMAC, both from the payment key. But it's fine to use the payment key for the HMAC directly, you just shouldn't use it for the encryption. It would probably be good to have a quick chat to figure out how we want the key derivation to work exactly.)